rust/osm2lanes/src/transform/tags_to_lanes/mod.rs

```diff
@@ -566,58 +568,75 @@ pub fn tags_to_lanes(
 }
 
 fn driving_lane_directions(tags: &Tags, _locale: &Locale, oneway: Oneway) -> (usize, usize) {
-    let both_ways = if let Some(n) = tags
-        .get("lanes:both_ways")
-        .and_then(|num| num.parse::<usize>().ok())
-    {
-        n
-    } else {
-        0
-    };
-    let num_driving_fwd = if let Some(n) = tags
+    let tagged_lanes = tags.get("lanes").and_then(|num| num.parse::<usize>().ok());
+    let tagged_forward = tags
         .get("lanes:forward")
-        .and_then(|num| num.parse::<usize>().ok())
-    {
-        n
-    } else if let Some(n) = tags.get("lanes").and_then(|num| num.parse::<usize>().ok()) {
-        let half = if oneway.into() {
-            n
-        } else {
-            // usize division rounded up
-            (n + 1) / 2
-        };
-        half - both_ways
-    } else if tags.is("lanes:bus", "2") {
-        2
-    } else {
-        1
-    };
-    let num_driving_back = if let Some(n) = tags
+        .and_then(|num| num.parse::<usize>().ok());
+    let tagged_backward = tags
         .get("lanes:backward")
-        .and_then(|num| num.parse::<usize>().ok())
-    {
-        n
-    } else if let Some(n) = tags.get("lanes").and_then(|num| num.parse::<usize>().ok()) {
-        let base = n - num_driving_fwd;
-        let half = if oneway.into() {
-            base
-        } else {
-            // lanes=1 but not oneway... what is this supposed to mean?
-            base.max(1)
-        };
-        half - both_ways
-    } else if tags.is("lanes:bus", "2") {
-        if oneway.into() {
-            1
-        } else {
-            2
+        .and_then(|num| num.parse::<usize>().ok());
+    let tagged_both_ways = tags
+        .get("lanes:both_ways")
+        .and_then(|num| num.parse::<usize>().ok());
+    let num_both_ways = tagged_both_ways.unwrap_or(0);
+    let tagged_bus_lanes = tags
+        .get("lanes:bus")
+        .and_then(|num| num.parse::<usize>().ok());
+
+    if oneway.into() {
+        // TODO support oneway=-1
+        if tagged_backward.is_some() {
+            warn!("lanes:backwards tag on a oneway road")
+        }
+        if tagged_both_ways.is_some() {
+            warn!("lanes:both_ways tag on a oneway road")
+        }
+        match (tagged_lanes, tagged_forward) {
+            (Some(l), None) => (l, 0),
+            (Some(l), Some(f)) => {
+                if l != f {
+                    warn!("lanes tag does not agree with lanes:forward on oneway road")
```

Comment:
we have a warnings Vec to structurally return this kind of information.
As we are a low level library, we need to give full control to the client on what they want to do with warnings

Reply:
This does seem like the right place to generate most of these warnings, so I will upgrade all of the `warn!`s to instead feed into the `RoadWarnings` vec.